ToDo REST API 만들기

BEOMKING opened the pull request:
안녕하세요.

Git pull request가 잘되는지 확인해보기 위해 커밋을 남겼습니다.

잘 부탁드립니다~

Commit:
Git pull request 세팅용 커밋

BEOMKING committed:
Todo의 주체인 Task 객체 생성

BEOMKING committed:
클라이언트와 통신할 HttpServer 생성

BEOMKING committed:
Merge branch 'BEOMKING' of https://github.com/BEOMKING/spring-week1-assignment-1 into BEOMKING

BEOMKING committed:
HttpHandler 생성 후 강의의 GET, POST 응답 메서드 작성

BEOMKING committed:
POST 요청시 Task 순서가 증가하며 생성

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DemoHttpHandler implements HttpHandler {
```
모든 객체에는 객체를 설명하는 주석을 추가해보는 것이 좋아요!

이 객체는 무엇을 하는지 편하게 주석을 작성해보아요!

BEOMKING replied:
넵. 객체의 역할에 관한 주석을 추가했습니다.
객체명도 변경해보았습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,69 @@
+        responseBody.close();
+    }
+
+    private Task toTask(String content) throws JsonProcessingException {
```
간단한 메소드에도 주석을 추가하면 나중에 많은 도움이 될 수 있어요!

BEOMKING replied:
질문있습니다!
메소드의 역할에 맞는 네이밍을 쓰는 것 이외에도 주석을 많이 활용하나요??

ssisksl77 replied:
안녕하세요! 
죄송해요... 질문을 잘 이해하지 못했어요! 
'메소드 역할 부가설명' 이외에 주석을 활용한다는 뜻일까요?

주석은 사실 '메소드 역할 부연설명' 만으로 바쁠거에요!

BEOMKING replied:
안녕하세요!!
질문을 더 잘하도록 노력해야겠어요.
메소드 네이밍에 많은 신경을 써서 작성하는 것 외에도
주석이 꼭 필요한지, 현업에서도 주석을 많이 활용하는지가 제 질문의 의미였습니다.

ssisksl77 replied:
현업에서 주석은 너무나도 중요합니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
     }
 
     public static void main(String[] args) {
+        try {
+            InetSocketAddress address = new InetSocketAddress(8080);
+            HttpServer httpServer = HttpServer.create(address, 0);
+            HttpHandler handler = new DemoHttpHandler();
+            httpServer.createContext("/", handler);
+            httpServer.start();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
예외가 실제로 던져지는 구간으로 try/catch를 좁혀보면 어떨까요?!

BEOMKING replied:
예외가 필요한 부분인 HttpServer만 따로 분리되도록 변경했습니다.

Commit:
객체가 하는 일과 관련되도록 네이밍 변경, 설명 주석 추가

Commit:
toTask의 역할에 대한 주석 추가

Commit:
HttpServer를 생성하는 로직 분리

Commit:
gitignore에서 out 하위 디렉터리 제외

Commit:
ObjectMapper.toString()이 주소값을 보내주어 String을 반환하는
writeValueAsString으로 변경

Commit:
요청온 경로가 유효하지 않으면 에러메시지를 반환하는 기능

Commit:
요청온 requestBody를 파싱해주는 기능, 요청을 처리할 분기문 생성

Commit:
불필요한 생성자 제거

BEOMKING committed:
Task 생성시 만든 Task를 확인할 수 있도록 Json으로 반환

BEOMKING committed:
path를 처리하기 위해 불필요한 조건문 제거

BEOMKING committed:
PUT 요청이 왔을 때, 요청한 Id의 Task를 변경하는 기능

BEOMKING reviewed (commented):
.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,100 @@
+    private String tasksToJson() throws IOException {
+        return objectMapper.writeValueAsString(tasks);
+    }
+    // 하나의 Task를 Json으로 반환해주는 메서드
```
```suggestion
  /**
   * Task를 Json String으로 변환하여 리턴합니다.
   * 
   * @param 
   * @throws
   **/
    // 하나의 Task를 Json으로 반환해주는 메서드
```

javaDoc 명세를 한번 따라해 보는 것이 좋은 도움이 될 거에요!
위에 `@param` `@throws` 도 한번 채워보아요!

BEOMKING replied:
JavaDoc으로 작성해봤습니다.
JavaDoc을 사용하면서 스스로 메서드의 역할, 파라미터, 예외, 리턴 값까지 한번 더 생각해볼 수 있어서 좋은 것 같아요.
다른 것들에도 적용해보는 습관을 가져보겠습니다.
[c20b65b](https://github.com/CodeSoom/spring-week1-assignment-1/pull/107/commits/c20b65bb8d6e3b8f9333400971bae250d3bd7e4d)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,100 @@
+        return path.equals("/tasks");
+    }
+
+    // 요청 받은 content를 Task 객체로 매핑하고 리턴
```
객체 매핑만 하고 있을까요? `toTask` 가 하는 다른일은 무엇이 있나요?

BEOMKING replied:
Task 객체로 매핑하고 Task의 Id를 set 해주고 있어요.
두 가지 일을 하고 있었네요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,100 @@
+public class TaskHttpHandler implements HttpHandler {
+    public static final String NOT_FOUND = "404 Not Found";
+    private List<Task> tasks = new ArrayList<>();
+    private ObjectMapper objectMapper = new ObjectMapper();
```
객체를 문자열로, 문자열을 객체로 바꿔주는 구현체가 꼭 TaskHttpHandler에 있어야 할까요? 
한번 분리하는 연습을 해보는 것은 어떨까요?!

BEOMKING committed:
Update app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java

Co-authored-by: 남영환 <ssisksl77@gmail.com>

BEOMKING committed:
taskToJson 메서드의 주석을 JavaDoc으로 작성

BEOMKING committed:
메서드들에 JavaDoc 추가

Commit:
PUT 요청한 Task의 Title을 변경해주는 로직을 추가했습니다.

Commit:
POST 처리 메소드 분리

Commit:
잘못된 요청을 받을 경우, 에러 코드와 에러 메시지를 반환하는 기능

Commit:
수신된 Http 요청이 잘못된 경우 응답 본문을 만드는 메서드로 변경

Commit:
handlePost 요청 본문 없는 경우 메시지 추가, JavaDoc 작성

Commit:
PUT 요청 처리 handlePut로 분리

Commit:
DELETE 요청시 경로의 id를 찾아 제거하는 로직을 작성했습니다.

Commit:
사용하지 않는 import 제거

BEOMKING committed:
TaskHttpHandler의 JavaDoc을 추가했습니다.

BEOMKING committed:
상세 조회 기능을 추가했습니다.

BEOMKING committed:
파라미터를 인라인화, 가독성이 더 좋다고 생각해서 변경

BEOMKING commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -7,15 +7,11 @@
-        HttpHandler handler = new TaskHttpHandler();
-
-        connectHttpServer(address, handler);
+        connectHttpServer(
+                new InetSocketAddress(8000),
+                new TaskHttpHandler());
```
생성하는 로직들을 인라인화하면 더 가독성이 좋아보여서 변경해봤습니다.
사용하지 말아야 할 이유나 발생할 수 있는 단점이 있을까요?

ssisksl77 replied:
아니요 너무 좋은데요?!

BEOMKING committed:
PUT, DELETE, GET 요청시 id가 유효하지 않을 때 404를 리턴하는 로직 추가

BEOMKING commented:
우선적으로 과제의 요구사항은 만족하였습니다.

BEOMKING committed:
Http 요청에 id가 포함되어 있는지 확인하는 로직을 메서드 분리

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -52,7 +52,10 @@ public void handle(HttpExchange exchange) throws IOException {
                         .orElseThrow(() -> new RuntimeException("찾을 수 없는 Task입니다."));
-                storedTask.setTitle("변경되었습니당.");
+
+                Task taskToChange = toTask(request);
+                storedTask.setTitle(taskToChange.getTitle());
```
Task를 저장하기 위해 꼭 Task 라는 객체로 변경 후에 제목을 가져와야 하는 걸까요?
request에서 곧바로 title을 사용하면 안될까요?

BEOMKING replied:
불필요한 작업인 것 같아요.
Map으로 매핑해서 바로 가져오도록 변경했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,241 @@
+
+        if (method.equals("GET")) {
+            if (isDetailMatches(path)) {
+                try {
+                    content = handleDetailGet(path);
+                    exchange.sendResponseHeaders(200, content.getBytes().length);
+                } catch (NoSuchElementException e) {
+                    exchange.sendResponseHeaders(404, -1);
+                }
+            } else {
+                content = tasksToJson();
+                exchange.sendResponseHeaders(200, content.getBytes().length);
+            }
+        } else if (method.equals("POST") && path.equals("/tasks")) {
+            content = handlePost(request);
+            exchange.sendResponseHeaders(201, content.getBytes().length);
+        } else if (method.equals("PUT") && isDetailMatches(path)) {
+            try {
+                content = handlePut(path, request);
+                exchange.sendResponseHeaders(200, content.getBytes().length);
+            } catch (NoSuchElementException e) {
+                exchange.sendResponseHeaders(404, -1);
+            }
+        } else if (method.equals("DELETE") && isDetailMatches(path)) {
+            try {
+                content = handleDelete(path);
+                exchange.sendResponseHeaders(204, -1);
+            } catch (NoSuchElementException e) {
+                exchange.sendResponseHeaders(404, -1);
+            }
+        } else {
+            content = handleBadRequest();
+            exchange.sendResponseHeaders(400, content.getBytes().length);
+        }
+
+        OutputStream responseBody = exchange.getResponseBody();
+        responseBody.write(content.getBytes());
+        responseBody.flush();
+        responseBody.close();
+    }
```
한가지 과재를 드릴게요! 
if문 안에 있는 모든 nested if문과 try-catch 문을 없애보는거에요!

BEOMKING replied:
🤔

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,241 @@
+    /**
+     * Task의 id가 경로에 포함되어 있으면 true, 아니면 false를 리턴합니다.
+     * @param path 수신된 Http 요청의 경로
+     * @return id가 경로에 포함되어 있으면 true, 아니면 fasle
```
`경로에 포함되어있다` 의 뜻은 `id가 숫자이다` 를 말씀하시는 걸까요?

BEOMKING replied:
id가 숫자인 경우라는 설명으로 변경해 의미를 좀 더 명확히 했습니다.

BEOMKING replied:
@ssisksl77 그런데 입력이 꼭 숫자 형식으로 온다고 확신할 수는 없을 것 같아요.
이런 경우, 현재는 400 코드를 보내주고 있는데 예외 처리를 추가로 진행해야 할까요? (예를 들면 입력이 문자인 경우)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,241 @@
+     * @throws NoSuchElementException 요청된 id를 찾지 못했을 경우 던집니다.
+     */
+    private String handleDetailGet(String path) throws JsonProcessingException, NoSuchElementException {
+        String content;
```
content라는 변수의 이름은 없어도 될 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    /**
+     * 특정 Task를 GET 요청이 왔을 때, 요청한 Task를 리턴한다.
+     * @param path 수신된 Http 요청의 경로
+     * @return 요청한 Task를 String으로 변환해서 리턴한다.
+     * @throws JsonProcessingException Json으로 변환하면서 에러가 발생한 경우 던집니다.
+     * @throws NoSuchElementException 요청된 id를 찾지 못했을 경우 던집니다.
```
```suggestion
     * 특정 Task를 GET 요청이 왔을 때, 요청한 Task를 리턴한다.
     * 
     * @param path 수신된 Http 요청의 경로
     * @return 요청한 Task를 String으로 변환해서 리턴한다.
     * @throws JsonProcessingException Json으로 변환하면서 에러가 발생한 경우 던집니다.
     * @throws NoSuchElementException 요청된 id를 찾지 못했을 경우 던집니다.
```
nit : 보통 javadoc 을 작성할 때 어노테이션이 나열되는 부분과 설명 사이에는 빈 한줄을 추가합니다.
아래 오라클 링크의 예시가 도움이 될 것 같아요!
https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html

BEOMKING replied:
이런 컨벤션이 있었군요. 감사합니다.

BEOMKING replied:
@ssisksl77 과제와 상관은 없지만 질문이 있습니다.
참고할 javadoc를 java8 docs에서 주셨는데 java18 docs에도 javadoc 가이드는 있습니다. 8을 보내주신 이유가 따로 있으신가요?
저는 공식문서를 볼 때 어느 버전의 doc를 참고하는게(찾는 정보가 두 버전에 모두 있을시에) 더 나을지 항상 혼란스러워 질문드립니다.

ssisksl77 replied:
오! 두 버전에 큰 차이가 없을텐데 어느부분이 혼란스러우셨나요?

ssisksl77 replied:
저는 일반적으로 공식문서를 볼 때 제가 사용하는 버전의 javadoc을 보게되는 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,241 @@
+        Long findId = Long.valueOf(splitedPath[2]);
+        String content;
+
+        if (!request.isBlank()) {
```
path를 쪼개기 전에 검사를 미리해서 빠른 실패를 만들 수 있지 않을까요?

BEOMKING replied:
먼저 쪼개서 값을 가지더라도 변경할 값을 가지고 있지 않으면 의미가 없네요!

BEOMKING committed:
요청을 Map으로 매핑해서 바로 title을 set 하도록 변경

Commit:
JavaDoc id 형식에 관한 설명 추가

Commit:
불필요한 변수명 제거

Commit:
JavaDoc 설명과 태그 컨벤션에 따라 한줄 띄우기

Commit:
path를 검사하기 전에 요청값이 먼저 있는지를 검사하도록 변경

Commit:
Task 관련 로직을 처리하는 TaskService 생성, Task를 생성해주는 로직 생성
및 테스트

Commit:
요청을 받아 Task를 생성하고 반환하는 로직

Commit:
Tasks 전체 조회 로직

Commit:
BadResponse 로직 추가

Commit:
특정 id를 GET 하는 요청을 받아 처리하는 로직

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,320 @@
+     * @return 빈 문자열을 리턴
+     * @throws NoSuchElementException 요청된 id를 찾지 못했을 경우 던집니다.
+     */
+    private String handleDelete(String path) {
+        String[] splitedPath = path.split("/");
+        Long findId = Long.valueOf(splitedPath[2]);
+
+        int index = -1;
+        for (int i = 0; i < tasks.size(); i++) {
+            if (tasks.get(i).getId().equals(findId)) {
+                index = i;
+                break;
+            }
+        }
+
+        if (index == -1) {
+            throw new NoSuchElementException();
+        }
+        tasks.remove(index);
+        return "";
+    }
```
```suggestion
    private String handleDelete(String path) {
        Long findId = Long.valueOf(path.split("/")[2]);

        for (int i = 0; i < tasks.size(); i++) {
            if (tasks.get(i).getId().equals(findId)) {
                tasks.remove(i);
                return i;
            }
        }
        throw new NoSuchElementException();
    }
```
이렇게 바꿀 수는 없을까요? 한번 확인부탁드려요!
splitedPath 를 제거한 이유는 코드 자체에서 `split` 하는 것을 알 수 있기 때문이에요!

BEOMKING replied:
코드 자체에서 `split`하는 것을 알면 어떤 문제가 있나요?

ssisksl77 replied:
뭔가 제가 글을 잘못 쓴 것 같군요! split을 분리해보자는 생각을 했던 것 같은데요 ㅠㅜㅠㅜ

일단 splitedPath를 제거한 이유는 한곳에서만 쓰기 때문입니다.

제가 원하는 코드는 `Long.valueOf(path.split("/")[2]);` 이 코드를 제거하는 것입니다.
`split` 을 알면 문제된다는 것은 `path` 에서 `id` 를 꺼내는 방식을 handleDelete 가 알아버린다는 것입니다!

ssisksl77 replied:
뭔가 제가 글을 잘못 쓴 것 같군요! split을 분리해보자는 생각을 했던 것 같은데요 ㅠㅜㅠㅜ

일단 splitedPath를 제거한 이유는 한곳에서만 쓰기 때문입니다.

제가 원하는 코드는 Long.valueOf(path.split("/")[2]); 이 코드를 제거하는 것입니다.
split 을 알면 문제된다는 것은 path 에서 id 를 꺼내는 방식을 handleDelete 가 알아버린다는 것입니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/service/TaskService.java:
```diff
@@ -0,0 +1,59 @@
+ * Task를 CRUD 로직을 처리하는 클래스입니다.
+ */
+public class TaskService {
+    private Long id = 0L;
+    private final List<Task> tasks = new ArrayList<>();
```
간단하게 병렬성을 고려한 코드를 만들어보면 어떨까요?

BEOMKING replied:
CopyOnWriteArrayList를 사용해 병렬성을 고려해보았습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/service/TaskServiceTest.java:
```diff
@@ -0,0 +1,32 @@
+        taskService.createTask("BJP");
+        assertEquals(new Task(0L, "BJP"), taskService.getTask(0L));
+    }
+}
```
EOF 에는 한 줄을 추가해주세요!

BEOMKING replied:
넵

BEOMKING replied:
개행이 없는 것이 파일에 예기치 않은 문제를 일으킬 수 있군요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/service/TaskServiceTest.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Test
+    @DisplayName("가진 tasks를 반환하는지 테스트한다.")
```
- 정확히 어떤 상황일 때
- 무엇이
- 무엇을 리턴하는지 
명시하는 것이 좋겠어요!

BEOMKING replied:
테스트 코드들에 적용했습니다. 좀 더 이해하기 쉬워진 것 같아요.

BEOMKING committed:
Update app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java

Co-authored-by: 남영환 <ssisksl77@gmail.com>

Commit:
테스트 코드 설명 구체적으로 변경

어떤 상황일 때 무엇이 무엇을 리턴하는지 다른 사람이 봤을 때 쉽게 알 수
있도록 변경했습니다.

Commit:
Merge branch 'BEOMKING' of https://github.com/BEOMKING/spring-week1-assignment-1 into BEOMKING

Commit:
PUT 요청을 받았을 때, 요청된 id에 맞는 Task를 수정하고 리턴하는 기능
추가

Commit:
response 일반화, 불필요한 코드 제거

응답 코드에 따라 생성되던 메서드를 응답 코드와 길이를 받도록
변경하였습니다.
if, try catch를 제거하면서 용도가 없어진 코드를 제거하였습니다.

BEOMKING commented:
코멘트에 label Pending이 있으면 리뷰어님이 볼 수가 없었군요..
이제 알았읍니다ㅠ

Commit:
요청된 id와 같은 Task를 제거해주는 로직, EOF 추가

Commit:
Handler에서 DELETE 요청을 처리하는 기능

Commit:
DELETE 요청시 잘못된 HTTP 코드 전달하는 것 수정, 불필요한 메서드 제거

BEOMKING commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -46,12 +46,22 @@ public void handle(HttpExchange exchange) throws IOException {
         sendResponse(exchange, 400, -1);
     }
 
+    private void sendDeleteResponse(HttpExchange exchange, String path) throws IOException {
+        try {
+            taskService.deleteTask(Long.valueOf(path.split("/")[2]));
+            sendResponse(exchange, 200, -1);
+        } catch (NoSuchElementException e) {
+            sendResponse(exchange, 400, -1);
+        }
+    }
```
여기서 try catch를 없앨 수 있는 방법이 뭐가 있을까요?
파일을 찾지 못했을 때, 어떤 결과를 보내주어야 할 지 잘 모르겠어요.

ssisksl77 replied:
그렇군요! try-catch를 없애고 싶으시군요! 
이 코드가 아닌 다른 관점으로 생각해볼게요!
어디에서 대체 예외를 던지고 있을까요? 

바로 deleteTask는 예외를 던지고 있고, 우리는 그 예외를 핸들링할 의무를 갖게 됩니다.
deleteTask는 파일을 찾기 못했을 때, 예외를 던짐으로써 핸들링할 책임을 미루고 있죠!

현재 몇가지 방법은 두가지네요
1.  true/false를 리턴한다
2. 삭제시 삭제한 객체를 리턴한다. 삭제를 하지않았으면 null을 리턴한다.
3. Either 라는 객체를 이용한다. (직접구현해서 사용할 수도 있습니다 https://www.baeldung.com/vavr-either)
1번과 2번은 자바 ArrayList 구현체에서도 사용하고 있어요! 
검증된 방법입니다.
https://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html#remove(int)

BEOMKING replied:
여러가지 방법이 있었네요. 1, 2번 방식으로 PUT, DELETE 구현을 변경했습니다.
null을 리턴하는 것도 괜찮은 방법이었군요!

ssisksl77 replied:
대신! 주석이 필요합니다! null을 아무렇게나 리턴하지 않음을 확실히 하는 것이 좋습니다.

ssisksl77 replied:
모던한 방식으로는 null을 완전히 배제하기 때문에 null은 편하고 고전적인 방식임을 인지해주시고 사용하시는 것이 좋아요!

Commit:
DELETE 요청시 HTTP 코드 잘못 주던 것 수정, 불필요한 메서드 제거

BEOMKING committed:
CopyOnWriteArrayList를 사용해 병렬성 고려

BEOMKING committed:
path를 나눠 id를 찾고 숫자로 변환하는 기능 메서드 분리

나누어진 path를 받아 Long을 변환하여 리턴하는 기능을 만들었습니다.

BEOMKING committed:
TaskMapper 생성

TaskHttpHandler가 처리하던 Mapper 역할을 TaskMapper에게 위임했습니다.

BEOMKING committed:
Http 응답 코드를 Enum클래스로 변경

ssisksl77 commented:
뭔가 이상하군요 제가 쓴 글들이 사라진 듯한 느낌이 듭니다. 제가 쓴 글에 답변만 적혀있는 느낌이 들어서 어떤 질문은 컨텍스트를 따라갈 수가 없네요.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,191 @@
+        URI uri = exchange.getRequestURI();
+        String path = uri.getPath();
+
+        if (method.equals("GET") && path.equals("/tasks")) {
+            sendGetResponse(exchange);
+            return;
+        }
+
+        if (method.equals("GET") && isDetailMatches(path)) {
+            sendDetailGetResponse(exchange, path);
+            return;
+        }
+
+        if (method.equals("POST") && path.equals("/tasks")) {
+            sendPostResponse(exchange);
+            return;
+        }
+
+        if (method.equals("PUT") && isDetailMatches(path)) {
+            sendPutResponse(exchange, path);
+            return;
+        }
+
+        if (method.equals("DELETE") && isDetailMatches(path)) {
+            sendDeleteResponse(exchange, path);
+            return;
+        }
```
왠지 라우팅 역할을 갖는 객체를 만들 수도 있지 않을까요?!!

BEOMKING replied:
Http 요청에 따라 다른 처리를 하는 Router들을 만들었습니다.
이 과정에서 Handler가 가지고 있는 역할이 많이 줄었습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskHttpHandler.java:
```diff
@@ -0,0 +1,191 @@
+            return;
+        }
+
+        HashMap requestMap = taskMapper.getRequestMap(request);
```
HashMap에 Key,Value 에 대한 타입을 적어주세요!

ssisksl77 replied:
일단 이것에 가능하려면 getRequestMap 이 개선되어야 하는군요!

ssisksl77 replied:
하지만 ObjectMapper를 이용하면 제네릭을 넣는 방법이 있을거에요!

일단 우리는 nested 한 해시맵이 없으니 테스트삼아 HashMap<String,String> 으로 변환하는 법을 연습해보면 어떨까요?

ssisksl77 commented:
https://github.com/CodeSoom/spring-week1-assignment-1/pull/107#discussion_r915324750

아하 400을 리턴해도 되냐는 말씀이시죠! 당연하죠! 잘못된 값을 받았으면 4XX 대로 리턴하면 됩니다!

BEOMKING committed:
try catch 제거

PUT 요청에서 변경된 Task를 리턴하거나 찾지 못했다면 null을 반환하는
방법으로 바꿨습니다.

DELETE 요청에서 삭제가 되었다면 true를 리턴하고 찾지 못했다면 false를
리턴해서 try catch를 제거했습니다.

BEOMKING committed:
Http 요청을 처리하는 Router 생성

Router를 생성하고 그 안에서 GET, POST, PUT, DELETE를 처리하는 Router를
만들었습니다.

BEOMKING committed:
objectMapper TypeReference로 타입 명시

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/mapper/TaskMapper.java:
```diff
@@ -0,0 +1,49 @@
+     * @throws JsonProcessingException Json 변환에 문제가 발생할 경우 던집니다.
+     */
+    public static HashMap<String, String> getRequestMap(String request) throws JsonProcessingException {
+        TypeReference<HashMap<String, String>> hashMapTypeReference = new TypeReference<>(){};
+        return objectMapper.readValue(request, hashMapTypeReference);
```
제가 원하는 것이었어요! 정말 좋습니다!
변수명을 지정해서 곧바로 사라지네요 인라인으로 만들어도 좋겠어요!
실무에서도 인라인으로 많이 사용해요!
```suggestion
        return objectMapper.readValue(request, new TypeReference<HashMap<String, String>>(){});
```

여기서 우리는 더 나아가야합니다.
이것이 왜 가능한 것인가! 
우리는 알아야합니다!

정리해서 공유해주세요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/router/Router.java:
```diff
@@ -0,0 +1,30 @@
+import java.io.IOException;
+
+public class Router {
+    private final TaskService taskService = new TaskService();
```
Router에 TaskService에 의존이 되면 좋지 않을 것 같아요!

BEOMKING replied:
이 의존이 어떤 문제가 되는지 계속 고민해봤는데 아직 잘 모르겠습니다 ㅠ

ssisksl77 replied:
아래 코멘트를 달았습니다. https://github.com/CodeSoom/spring-week1-assignment-1/pull/107#discussion_r917391842

BEOMKING committed:
변수 인라인

BEOMKING committed:
Router TaskService에 대한 의존성 제거..하려 노력

BEOMKING commented on app/src/main/java/com/codesoom/assignment/router/Router.java:
```diff
@@ -6,11 +6,19 @@
-    private final PutRouter putRouter = new PutRouter(taskService);
-    private final DeleteRouter deleteRouter = new DeleteRouter(taskService);
-    private final GetRouter getRouter = new GetRouter(taskService);
+    private final TaskService taskService;
+    private final PostRouter postRouter;
+    private final PutRouter putRouter;
+    private final DeleteRouter deleteRouter;
+    private final GetRouter getRouter;
+
+    public Router(TaskService taskService) {
+        this.taskService = taskService;
+        postRouter = new PostRouter(taskService);
+        putRouter = new PutRouter(taskService);
+        deleteRouter = new DeleteRouter(taskService);
+        getRouter = new GetRouter(taskService);
+    }
```
이 방법은 결국 Handler에게 taskService를 주입 받고 다른 라우터에게 전해주는 방법이라 이전과 크게 다를 것이 없는 것 같아요.
어떤 방법으로 해결할 수 있었을까요? 너무 궁금합니다

ssisksl77 replied:
아하! Router라는 것에 TaskService가 아닌 다른 Service가 추가될 때마다 Router가 추가되어야 한다면!
`Router` 는 GodObject가 될 확률이 높아보여요!

ssisksl77 replied:
어쩌면 이런 코드가 시작이 되지 않을까 싶습니다. 

```java
new Router(
 new PostRouter(taskService), 
 new PutRouter(taskService), 
 new DeleteRouter(taskService), 
 new GetRouter(taskService));
```

ssisksl77 commented:
한 주간 고생하셨습니다! 다음 주도 잘 부탁드려요!
